### ISSUE TITLE
Unexpected blocking `KEYS` command at 2.0.0

### ISSUE THREAD

**Issue by viralpraxis** [sidekiq-cron/sidekiq-cron#516]
We upgraded `sidekiq-cron` to version `2.0.0` from `1.12.0`, and noticed that at Sidekiq's startup, it performs a blocking `KEYS cron_jobs:*` command. This command takes over 5 seconds on large Redis databases, which leads to client timeouts.

Before:

```
$ redis-cli monitor | grep 'cron_job'
1731416178.107613 [1 [::1]:41712] "smembers" "cron_jobs"
```

After:
```
$ redis-cli monitor | grep 'cron_job'
1731416133.431067 [1 [::1]:50116] "keys" "cron_jobs:*"
1731416133.431376 [1 [::1]:50116] "smembers" "cron_jobs:default"
```

We have two questions regarding this behavior:

1. We reviewed the changelog but didn’t see any mention of this change. Local and production-like testing didn’t reveal the problem, as it only occurs with larger Redis installations (over 10 million keys). Should this be highlighted somewhere, especially considering there's a [dedicated warning regarding using this command in production](https://redis.io/docs/latest/commands/keys/)?.
2. Is the use of the KEYS wildcard truly necessary?

**Comment by markets** [comment]
Hello @viralpraxis, 

The bigger change in v2 is the introduction of Namespaces, see the whole discussion: https://github.com/sidekiq-cron/sidekiq-cron/pull/268. I'm quite sure this could be related: 

https://github.com/sidekiq-cron/sidekiq-cron/blob/b93aec83d3a8c31834b071d4da61f4f48cf97aba/lib/sidekiq/cron/namespace.rb#L8

But I didn't noticed any problem in my apps running Sidekiq-Cron v2. Also, we released a couple of release candidates and nobody reported any related problem.

**Comment by viralpraxis** [comment]
Hello @markets,

Yep, seems like that's the only place in gem where `KEYS` is performed.

> But I didn't noticed any problem in my apps running Sidekiq-Cron v2

Yep, the issue only occurs on relatively big databases (we have ~40M keys) with small enough client read timeout (we have 5s). Iterating takes 6 seconds on our production machines (which is expected AFAIK).

**Comment by viralpraxis** [comment]
There are two steps we can take to mitigate the issue. First, while I completely understand that the `namespaces = conn.keys('cron_jobs:*').collect do |key|` part was introduced in version 2.0.0 as an implementation detail, it has a notable side effect of blocking the entire Redis server for a duration proportional to the number of keys. I believe we are not the only Sidekiq-Cron users with a database of this size (which I wouldn’t consider "extremely huge", just relatively big), so I think it should be mentioned that the gem now performs a full database scan at startup (maybe in release description?). The issue is hard (if not impossible) to catch in staging environments for obvious reasons.

Secondly, it might be possible to replace the `KEYS` command with `SCAN`, which would alleviate some of the blocking issues. I'm not familiar with the codebase, so I'm not yet sure if this approach is applicable here, but I'll take a look.

**Comment by viralpraxis** [comment]
I've looked into it, and it seems that using `SCAN` instead of KEYS offers limited benefit. While `SCAN` prevents Sidekiq from blocking the entire Redis server for [many] seconds, each call to `Sidekiq::Cron::Namespace.{all,all_with_count}` would still take considerable time (basically time for `KEYS` plus extra latency overhead), making it practically unusable.

Note that Sidekiq itself explicitly avoids using this type of query (`O(n)`, where `n` is **total** keys count) by working with queues specified at configuration time. Iterating over jobs within a specific queue only requires a single lookup in Redis's queue structure.

ATM we will probably just monkeypatch this method like this

```ruby
class Sidekiq::Cron::Namespace
  def self.all
    [Sidekiq::Cron.configuration.default_namespace]
  end
end
```

but it would be cool to make the entire "namespacing" optional.

**Comment by forsbergplustwo** [comment]
We ran into this as well today, locked everything up and brought down production. We have 10m+ keys, and a connect timeout of 3 seconds, which seems not enough for this lookup.

**Comment by markets** [comment]
@zedtux as the main developer of the _Namespaces_ (#268) feature, I'd really appreciate 🙏🏼  your thoughts on this issue.

**Comment by zedtux** [comment]
We [discussed that point](https://github.com/sidekiq-cron/sidekiq-cron/pull/268#issuecomment-1114835623) while I was working on the feature, but we missed the performance point.

> it would be cool to make the entire "namespacing" optional

It's not necessary to make the entire "namespacing" optional while the migration code is the only painful point for you, especially that now namespacing is deeply implemented in this gem's code.

Instead I would look for one of those options I could think of so far :
* Update jobs when working on them, in order to keep this kind of "transparent" mode
* Move the migration code in a task allowing one to run it in a separated job (sidekiq?)

Regarding the second point, it should be checked if there could be issues to run non-migrated jobs.

Let's find the best way to tackle this issue!

**Comment by markets** [comment]
> It's not necessary to make the entire "namespacing" optional while the migration code is the only painful point for you, especially that now namespacing is deeply implemented in this gem's code.

I agree, at this point, this doesn't seem like an option. 

I still think we need to keep the "transparent" migration path, as having a separate task needs also coordination between the deploy and the script. I migrated like 4 apps to this version with no problems at all, but I don't have such big Redis sets.

But if I understood well the problem, the issue occurs when calling:

https://github.com/sidekiq-cron/sidekiq-cron/blob/b93aec83d3a8c31834b071d4da61f4f48cf97aba/lib/sidekiq/cron/namespace.rb#L8

So this will keep happening even if we change the migration code, no?

**Comment by zedtux** [comment]
With the "Update jobs when working on them" option, the idea is to check the current job, if it's a migrated job or not, when calling some key methods of this gem.

When the job is not migrated, migrate it and then let the code continue.

**Comment by markets** [comment]
But, please correct me if I'm wrong, the issue occurs (blocking the entire Redis) when calling `Sidekiq::Cron::Namespace.all`, so even if we change the migration strategy, the issue will occur.

**Comment by viralpraxis** [comment]
I can confirm that `KEYS cron_jobs:*` is performed each time `/sidekiq/cron` is accessed.

**Comment by zedtux** [comment]
Oh okay! You're right, the jobs migration has nothing to do here.

What I can suggest then is to build a `namespaces` table that would be filled once at startup time, and then use that table instead of scanning the entire jobs table.

What do you think about that ?

**Comment by markets** [comment]
Yes, I think we have 2 options:
- New structure to save available "namespaces" in Redis (as @zedtux [pointed out above](https://github.com/sidekiq-cron/sidekiq-cron/issues/516#issuecomment-2482442804)).
- Make it like a configuration option, ala `available_namespaces`.

Not sure which one is better 😅

**Comment by zedtux** [comment]
Definitely the first option to me. Let's see what others vote for.

**Comment by viralpraxis** [comment]
I believe the `KEYS` command should be avoided entirely. It’s comparable to Rails performing a full sequential scan (without an index) across an entire relational database. While it might work fine in a local environment or small production setups, it would be a disaster in real-world systems. With Redis, the situation is even worse because the command blocks the entire server.

Scanning Redis once (e.g., during startup) is also problematic, as it still blocks the server for seconds and could prevent other clients or gems from initializing. In the best-case scenario, replacing `KEYS` with batched `SCAN` could result in Sidekiq being delayed by ~10–inf seconds, depending on the size of the Redis database.

Option 2 aligns with how Sidekiq’s queues work, and everything I’ve mentioned explains why it’s implemented this way.

**Comment by viralpraxis** [comment]
The third option is to retain the current behavior if `available_namespaces` is not specified (or e.g. set to special value `:automatic`). This approach would satisfy both users who want it to work automatically and those of us managing large Redis deployments.

**Comment by viralpraxis** [comment]
> Sure, we could go that way, but what about a job which is created with a namespace not included in this list? Should it be rejected?

Can we just emit a warning and assign the "default" namespace?

> What is "the current behavior"?

I meant pre-v2 behavior, without `KEYS` query.

**Comment by markets** [comment]
If we take this path (moving it to configuration option), I think we should keep anyway the _namespacing_ feature always on, by using the "default" as the default available namespace.

It also remains me how the I18n module works and its `available_locales` setting.

**Comment by Yurokle** [comment]
Hello. Just wanted to +1 and mention that my company is also seeing this issue. We have a dozens of ruby services which use `sidekiq-cron` gem and `keys` command is also forbidden by our database ops team for a good reason - as was already mentioned above. So we can’t upgrade to the versions 2 and above.

Thank you so much for working on the fix!

**Comment by zedtux** [comment]
> Can we just emit a warning and assign the "default" namespace?

That sounds good to me.

So to recap, this gem need a `available_namespaces` configuration option that:
* when is not defined would keep the current behavior (meaning using the Redis `KEYS` method)
* when defined would be used to return the list of namespaces AND warn when a job is created with a namespace which is not included in that list, but use the default namespace

**Comment by markets** [comment]
@zedtux I'm totally fine with the latest proposal (https://github.com/sidekiq-cron/sidekiq-cron/issues/516#issuecomment-2490397864). It seems that @viralpraxis is also ok with that approach. Would you like to work on a PR?

**Comment by zedtux** [comment]
I would love to, but i don’t have the time now, and until end of Feb, sorry.

**Comment by markets** [comment]
No prob @zedtux! Fortunately @viralpraxis already sent a patch. Could you please give us your feedback?

**Comment by eileen-jiang** [comment]
Hi, we're upgrading from < v2 and not planning on using the namespaces feature.

To avoid any `KEYS` commands, is this the correct setup?
```
Sidekiq::Cron.configure do |config|
  config.available_namespaces = [Sidekiq::Cron.configuration.default_namespace]
end
```

**Comment by viralpraxis** [comment]
Hey @eileen-jiang, that's correct.

**Comment by taylorrf** [comment]
@Pavel-Guseynov It's possible to share more details about the issue in production you had?

Just asking because I planning to upgrade from `1.12.0` to `2.2.0` adding this initializer config:
```
Sidekiq::Cron.configure do |config|
  config.available_namespaces = [Sidekiq::Cron.configuration.default_namespace]
end
```

Did you used some similar setup and still get issues?

**Comment by viralpraxis** [comment]
Honestly, I agree with you that namespacing should be optional and disabled by default. We can achieve this simply by setting `available_namespaces` to `%w[default]` instead of `nil` on the gem's side. The main issue right now is that it's not entirely clear to users that using `sidekiq-cron` without explicitly configured namespaces can break apps with large Redis deployments—and that's very wrong, in my opinion.

**Comment by viralpraxis** [comment]
> I have not dug into the details of the namespaces feature, but my understanding is that if it is not explicitly configured it will fallback to the default namespace. The same default namespace is configured with available_namespaces and nothing will break. Or will it?

Nope, if you do not configure namespaces explicitly (i.e. in your `sidekiq.rb` initializer) it will be `nil` by default and `sidekiq-cron` will use the `KEYS` command to collect the data about existing namespaces. That's why you (and me, and some other people in this thread) had troubles in production.

> Either way, this is still just a workaround and the main fix should be not to use the keys command at all. From my perspective, namespaces can't be used now in high-load environments without changin that.

I agree. But instead of using `SCAN`, I believe the best solution is to follow Sidekiq's approach and do the same as they did with queues: always require the user to specify existing namespaces via a configuration option (`available_namespaces` in our case).

To sum up, I completely agree with your point, but it's obviously not up to me to decide. @markets, could you please share your thoughts?

**Comment by markets** [comment]
I'm totally ok to improve the current situation, I'm currently a bit busy with a lot of stuff and I'm not able to work on that, but please, feel free to explore a solution (ideally with backwards compatibility 🙏).

**Comment by viralpraxis** [comment]
I don't really see how we can fix the things and keep total backwards compatibility. Given that we all agree that performing `KEYS` by default is unacceptable here's the best way to manage this issue I can think of:

1. Change default value of `available_namespaces` from `nil` to `%w[default]`
2. Add support value `available_namespaces = :auto` which will infer available environments using `KEYS`

It will be backward-incompatible only for people using namespacing and not specifying them explicitly -- they'll have to either specify `:auto` (to keep current behaviour) or list all the used namespaces by hand; new users and users not using namespacing will be fine.
